### ISSUE TITLE
Garden project scanning slow, causing OOMs

### ISSUE THREAD

**Issue by swist** [garden-io/garden#1574]
## Bug
We are working on moving a large-ish monorepo to garden. Our frontend packages use lerna, which means our node_modules are quite large. We have included `node_modules` in our `.gitignore`
### Current Behavior
Garden cli OOMs when we try to run `garden scan` or `garden dev`. 
```
garden scan
✔ providers                 → Getting status... → Done

<--- Last few GCs --->
in[46621:0x108000000]    46535 ms: Mark-sweep 1901.3 (1962.6) -> 1900.1 (1961.6) MB, 82.1 / 2.0 ms  (+ 1489.9 ms in 341 steps since start of marking, biggest step 13.4 ms, walltime since start of marking 1597 ms) (average mu = 0.225, current mu = 0.192) fin[46621:0x108000000]    47992 ms: Mark-sweep 1900.8 (1961.6) -> 1899.7 (1961.1) MB, 52.0 / 2.0 ms  (+ 1379.9 ms in 325 steps since start of marking, biggest step 14.2 ms, walltime since start of marking 1450 ms) (average mu = 0.210, current mu = 0.193) fin

<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x10092aa79]
Security context: 0x2271b10808a1 <JSObject>
    1: read [0x2271fef63269] [pkg/prelude/bootstrap.js:~563] [pc=0x225aff23519f](this=0x2271c5942849 <Object map = 0x22711fd39af1>,10710,0x227176eeac11 <Uint8Array map = 0x2271f1e493f1>,0,65536,0x22718ad004a9 <undefined>)
    2: arguments adaptor frame: 6->5
    3: _read [0x22717e6a2561] [internal/fs/streams.js:~132] [pc=0x225aff3bd002](this=0x2271d2d79389 <Rea...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x100faf490 node::Abort() (.cold.1) [/usr/local/bin/garden]
 2: 0x10007f43c node::FatalError(char const*, char const*) [/usr/local/bin/garden]
 3: 0x10007f564 node::OnFatalError(char const*, char const*) [/usr/local/bin/garden]
 4: 0x1001454c7 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/garden]
 5: 0x100145467 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/garden]
 6: 0x1002d0635 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/usr/local/bin/garden]
 7: 0x1002d1eaa v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/usr/local/bin/garden]
 8: 0x1002ce97c v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/usr/local/bin/garden]
 9: 0x1002cc79e v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/garden]
10: 0x1002cb7e1 v8::internal::Heap::HandleGCRequest() [/usr/local/bin/garden]
11: 0x1002903fe v8::internal::StackGuard::HandleInterrupts() [/usr/local/bin/garden]
12: 0x1005ebc4c v8::internal::Runtime_StackGuard(int, unsigned long*, v8::internal::Isolate*) [/usr/local/bin/garden]
13: 0x10092aa79 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/usr/local/bin/garden]
Abort trap: 6
```

### Expected behavior
Garden.io does not OOM

### Reproducible example

Can't really provide one, our repo looks more or less like this:
```
some_infra
  garden.yml
some_other_infra
  garden.yml
services/
  service1/garden.yml
  service2/garden.yml
  service3/garden.yml
  service4/garden.yml
  service5/garden.yml
  service6/garden.yml
  service7/garden.yml
  service8/garden.yml
frontend
  garden.yml
  lerna.json
  package.json
  package-lock.json
  node_modules (really huge)
```

### Workaround
Garden does not OOM if we add a `.gardenignore` next to our largest `node_modules`

### Suggested solution(s)
I don't know if garden tries to recurse all the way down the directory tree to find the service configuration, but perhaps we could specify a glob that lets it find the garden.ymls? Something like `node-glob` style matches or `minimatch` style matches? If you point me to the place where the file scanning takes place I'm more than happy to experiment against our repo

### Additional context
It also seems that the discovery is quite slow, see below:
```
$ time garden scan
✔ providers                 → Getting status... → Done
real	1m5.312s
user	2m21.535s
sys	0m21.655s
```

Here's a link to a slack discussion:
https://kubernetes.slack.com/archives/CKM7CP8P9/p1581017225194700
My colleague has posted a spindump

### Your environment

We're on OS X catalina, using docker-desktop as our local k8s provider. This seems to happen before any k8s work though.

```
garden version
0.11.3
```

```
kubectl version
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T23:42:50Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.5", GitCommit:"20c265fef0741dd71a66480e35bd69f18351daea", GitTreeState:"clean", BuildDate:"2019-10-15T19:07:57Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
```

```
docker version
Client: Docker Engine - Community
 Version:           19.03.5
 API version:       1.40
 Go version:        go1.12.12
 Git commit:        633a0ea
 Built:             Wed Nov 13 07:22:34 2019
 OS/Arch:           darwin/amd64
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          19.03.5
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.12
  Git commit:       633a0ea
  Built:            Wed Nov 13 07:29:19 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Comment by swist** [comment]
I think this might be a duplicate of https://github.com/garden-io/garden/issues/1553

**Comment by edvald** [comment]
Thanks for the detailed report @swist! This is unrelated to #1553 I think.

Could you add your project `garden.yml` to the mix? Minus anything sensitive of course. Then we have everything we need to explore, I think.

**Comment by edvald** [comment]
One other thing to clarify: Is it the `Resolving providers...` step that's taking this long? That actually doesn't relate to the directory scanning, and would point to something different relating to the Kubernetes provider.

**Comment by swist** [comment]
It's `garden scan` that takes a long time


```
kind: Project
name: marmot
environments:
  - name: local
    variables: &variables
  - name: remote
    variables:
      <<: *variables
dotIgnoreFiles:
  - .gardenignore
providers:
  - name: local-kubernetes
    environments: ["local"]
    setupIngressController: nginx
    defaultHostname: marmot.local
    namespace: garden-${local.env.ENV_NAME}
    tlsCertificates:
      - name: marmot-local
        secretRef:
          name: tls-marmot-local
          namespace: default
          
```

But looking at the root one, it seems like we have explicitly opted out of using `.gitignore` to ignore files to scan, let me try if that is the reason we're moving so much stuff across!

**Comment by swist** [comment]
Yep adding `.gitignore` to `dotIgnoreFiles` has fixed it, and it takes  ~20s to complete `garden scan`

**Comment by edvald** [comment]
Ah, that does explain it then! Actually, `garden scan` currently also resolves providers in order to be able to fully resolve modules, which involves connecting to the `kubernetes` cluster and checking some system services. That's kind of a bug but fixing it properly unfortunately requires a breaking change, so we've slated that for 0.12.

**Comment by 10ko** [comment]
Hi @edvald, do we have an issue for the `garden scan` taking long?
If so I'd close this in favour of that.

**Comment by edvald** [comment]
I don't think there is one already

**Comment by swist** [comment]
We're pretty sure this is fixed as of 0.11.11

**Comment by eysi09** [comment]
I'll go ahead and close this then.
